daffodil-lib/src/main/scala/edu/illinois/ncsa/daffodil/util/DecimalUtils.scala

```diff
@@ -0,0 +1,223 @@
+ * 
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal with
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is furnished to do
+ * so, subject to the following conditions:
+ * 
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimers.
+ * 
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimers in the
+ *     documentation and/or other materials provided with the distribution.
+ * 
+ *  3. Neither the names of Tresys Technology, nor the names of its contributors
+ *     may be used to endorse or promote products derived from this Software
+ *     without specific prior written permission.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE
+ * SOFTWARE.
+ */
+
+package edu.illinois.ncsa.daffodil.util
+
+import com.ibm.as400.access.AS400PackedDecimal
+
+import java.math.{ BigInteger => JBigInteger, BigDecimal => JBigDecimal }
+
+object DecimalUtils {
+
+  def packedToBigDecimal(packedNum: Array[Byte], scale: Int): JBigDecimal = {
+    val packed: AS400PackedDecimal = new AS400PackedDecimal((packedNum.length * 2) - 1, scale)
+
+    packed.toObject(packedNum).asInstanceOf[JBigDecimal]
+  }
+
+  def packedFromBigDecimal(num: JBigDecimal, nBits: Int): Array[Byte] = {
+    val packed: AS400PackedDecimal = new AS400PackedDecimal(nBits/4 - 1, num.scale)
+
+    packed.toBytes(num)
+  }
+
+  def bcdToBigInteger(bcdNum: Array[Byte]): JBigInteger = {
+    val numDigits: Int = bcdNum.size * 2  // 2 digits stored per byte
+    val outputData: Array[Char] = new Array[Char](numDigits)
+    var outputPos: Int = 0
+    var offset: Int = 0
+    var nibble: Int = 0
+
+    while (offset < bcdNum.size) {
+      nibble = (bcdNum(offset) & 0xFF) >>> 4
+      if (nibble > 0x09)
+        throw new NumberFormatException("Invalid high nibble")
+
+      outputData(outputPos) = (nibble | 0x0030).toChar
+      outputPos = outputPos + 1
+
+      nibble = (bcdNum(offset) & 0x0F)
+      offset = offset + 1
+      if (nibble > 0x09)
+        throw new NumberFormatException("Invalid low nibble")
+
+      outputData(outputPos) = (nibble | 0x0030).toChar
+      outputPos = outputPos + 1
+    }
+
```

Comment:
Instead of building up a String, I'm wondering if it would be more efficient to instead to just build up a number. So for each new nibbled, something like
```
value = 0
while (...) {
  nibble = // get new nibble
  value = value * 10 + nibble
}
```
This would avoid allocating a char array, convert to a String, and then parse it into a JBigInteger. Should be a bit more efficient.